builds/jazz-build-module/apigee-module.groovy

```diff
@@ -0,0 +1,285 @@
+ * @param env_key The type of deployment. e.g. PROD or TEST.
+ * @param environment_logical_id The id used to differentiate proxies from different branches in the same domain.
+ * @param config The config metadata.
+ */
+def delete(swaggerFile, env_key, environment_logical_id, config) {
+    def templateValues = getTemplateValues(swaggerFile, environment_logical_id)    
+    def deployEnv = apigeeConfig.API_ENDPOINTS[env_key]
+    dir (apigeeModuleRoot) {
+        withCredentials([usernamePassword(credentialsId: apigeeConfig.apigeeCredId, passwordVariable: 'PASS', usernameVariable: 'USER')]) {
+            try {
+                events.sendStartedEvent('APIGEE_API_PROXY_DELETE', 'Deleting Apigee API proxy.')
+                sh("./build/delete.sh ${deployEnv.mgmt_host} ${deployEnv.mgmt_org} ${deployEnv.mgmt_env} ${templateValues.ProxyName} ${USER} ${PASS}")
+                events.sendCompletedEvent('APIGEE_API_PROXY_DELETE', 'Completed deletion of Apigee API proxy.')
+            } catch (e) {
+                events.sendFailureEvent('APIGEE_API_PROXY_DELETE', e.getMessage())
+                throw e
+            }
+        }
+    }
+}
+
+/**
+ * Run the proxygen process to generate the api proxy definition files.
+ *
+ * @param templateValues A Map of token name, replacement values to use when filling templates.
+ */
+def proxygen(templateValues) {
+    dir ('gen') {
+        deleteDir()
+    }    
+    dir ('templates/CoreAPI') {
+        deleteDir()
+    }
+    dir ('templates') {
+        sh "mkdir ./CoreAPI"
+        if (apigeeConfig.useSecure) {
+            sh ("cp -r ./coreAPI_secure/* ./CoreAPI")
+        } else {
+            sh ("cp -r ./coreAPI_default/* ./CoreAPI")
+        }
+    }
+
+    applyTemplate('proxygen_template.properties', '.', 'proxygen.properties', templateValues)
+    sh ('cat proxygen.properties')
+    sh ("java -cp ./ProxyGen.jar com.tmobile.apigee.ProxyCodeGen proxygen.properties")
+    sh ("mv ./gen/CoreAPI/Proxies/* ./gen/CoreAPI")
+    sh ("mkdir ./gen/CoreAPI/build")
+    sh ("cp ./build/* ./gen/CoreAPI/build")
+}
+
+/**
+ * Construct the url of the newly created API proxy.
+ *
+ * @param deployEnv The deployment env used for this deployment.
+ * @param swaggerFile The swagger file for this deployment.
+ * @return String with the url for the deployed API proxy.
+ */
+String getHostUrl(deployEnv, swaggerFile) {
+    def swaggerFileContent = readFile swaggerFile
+    def swaggerObj = new JsonSlurperClassic().parseText(swaggerFileContent)
+    def domain = swaggerObj.basePath.substring(1)
+    def serviceName
+    for (k in swaggerObj.paths.keySet()) {        
+        def keyTokens = k.split('/')
+        if (keyTokens.length < 2) {
+            error "Path must contain at least one element after the initial / to indicate the service name."
+        }
+        serviceName = keyTokens[1]
+        break;
+    }
+    return "https://${deployEnv.service_hostname}/${domain}/${serviceName}"
```

Comment:
I see that we use environment_logical_id during proxy creation but the Url here isn't using it. Not using environment_logical_id will break the multi-environment feature. 

For example, if two dev environments are created, as per above line, they share the same url for their services (which is incorrect, they should have their own service endpoint).

Reply:
The domain value is pulled from the basepath which includes the environment_logical_id for DEV deployments. (This is done in the build_pack_api Jenkins file when it is updating the swagger file.)

Reply:
Got it. Thanks!